src/omnicore/rules.cpp

```diff
@@ -322,48 +338,66 @@ bool ActivateFeature(uint16_t featureId, int activationBlock, int transactionBlo
         case FEATURE_BETTING:
+            if (params.MSC_BET_BLOCK <= transactionBlock) {
+                PrintToLog("Feature activation of ID %d refused as the feature is already live\n", featureId);
+                return false;
+            }
             MutableConsensusParams().MSC_BET_BLOCK = activationBlock;
-            PrintToLog("Feature activation of ID %d succeeded. "
-                       "Bet transactions are going to be enabled at block %d.\n",
-                       featureId, params.MSC_BET_BLOCK);
-            return true;
+            featureName = "Bet transactions";
+        break;
         case FEATURE_GRANTEFFECTS:
+            if (params.GRANTEFFECTS_FEATURE_BLOCK <= transactionBlock) {
+                PrintToLog("Feature activation of ID %d refused as the feature is already live\n", featureId);
+                return false;
+            }
             MutableConsensusParams().GRANTEFFECTS_FEATURE_BLOCK = activationBlock;
-            PrintToLog("Feature activation of ID %d succeeded. "
-                       "The potential side effect of crowdsale participations, when "
-                       "granting tokens, is going to be disabled at block %d.\n",
-                       featureId, params.GRANTEFFECTS_FEATURE_BLOCK);
-            return true;
+            featureName = "Remove grant side effects";
+        break;
         case FEATURE_DEXMATH:
             MutableConsensusParams().DEXMATH_FEATURE_BLOCK = activationBlock;
-            PrintToLog("Feature activation of ID %d succeeded. "
-                       "Offering more tokens than available on the traditional DEx "
-                       "is no longer allowed and going to be disabled at block %d.\n",
-                       featureId, params.DEXMATH_FEATURE_BLOCK);
-            return true;
+            featureName = "DEx integer math update";
+        break;
+        default:
+            featureName = "Unknown feature";
+            supported = false;
+        break;
     }
-
-    PrintToLog("Feature activation of id %d refused due to unknown feature ID\n", featureId);
-    return false;
+    PrintToLog("Feature activation of ID %d processed. %s will be enabled at block %d.\n", featureId, featureName, activationBlock);
+    DeletePendingActivation(featureId); // if this feature id was previously scheduled for activation, delete the stale pending object
+    AddPendingActivation(featureId, activationBlock, minClientVersion, featureName);
+    if (!supported) {
+        PrintToLog("WARNING!!! AS OF BLOCK %d THIS CLIENT WILL BE OUT OF CONSENSUS AND WILL AUTOMATICALLY SHUTDOWN.\n", activationBlock);
+        std::string alertText = strprintf("Your client must be updated and will shutdown at block %d (unsupported feature %d ('%s') activated)\n",
+                                          activationBlock, featureId, featureName);
+        AddAlert("omnicore", ALERT_BLOCK_EXPIRY, activationBlock, alertText);
+        CAlert::Notify(alertText, true);
```

Comment:
Hmm. We fire the other notification in `logicMath`. I'm wondering, whether we could/should move `CAlert::Notify` into `AddAlert`, ideally including the log output. Would this make sense? If so, is it possible, or would this trigger notifications, which shouldn't be triggered? (e.g. when loading alerts or so)